SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularitySlaveAndRackManager.java

```diff
@@ -208,13 +210,23 @@ public SlaveMatchState doesOfferMatch(SingularityOfferHolder offerHolder, Singul
         }
         break;
       case OPTIMISTIC:
-        final double numPerSlave = numDesiredInstances / (double) stateCache.getNumActiveSlaves();
+        Collection<SingularityTaskId> currentlyActiveTasksForRequestClusterwide = stateCache.getActiveTaskIdsForRequest(taskRequest.getRequest().getId());
 
-        final boolean isSlaveOk = numOnSlave < numPerSlave;
+        // If no tasks are active for this request yet, we can fall back to greedy.
+        if (currentlyActiveTasksForRequestClusterwide.size() > 0) {
 
-        if (!isSlaveOk) {
-          LOG.trace("Rejecting OPTIMISTIC task {} from slave {} ({}) due to numOnSlave {}", taskRequest.getRequest().getId(), slaveId, host, numOnSlave);
-          return SlaveMatchState.SLAVE_SATURATED;
+          Set<String> currentHostsForRequest = currentlyActiveTasksForRequestClusterwide.stream()
+              .map(SingularityTaskId::getSanitizedHost)
+              .collect(Collectors.toSet());
+
+          final double numPerSlave = currentlyActiveTasksForRequestClusterwide.size() / (double) currentHostsForRequest.size();
+
+          final boolean isSlaveOk = numOnSlave <= numPerSlave;
+
+          if (!isSlaveOk) {
+            LOG.trace("Rejecting OPTIMISTIC task {} from slave {} ({}) due to numOnSlave {}", taskRequest.getRequest().getId(), slaveId, host, numOnSlave);
```

Comment:
might want to add the other values you're using in he calculation in here for debugging purposes